### PR TITLE
Creates a vectorized version of ex2

### PR DESCRIPTION
**Pull request by bishtgautam**
The RHSFunction is split into functions for internal and boundary edges. Additionally,
following two functions are vectorized:
- `GetVelocityFromMomentum`: (h,hu,hv) --> (u,v)
- `Solver`: The Roe solver

### CODE REVIEW

**Comment by bishtgautam** [comment]
@jeff-cohere @donghuix I took a crack at vectorizing the example based on our discussion this morning.

File: swe_roe/ex2b.c [hunk 1]
```diff
@@ -0,0 +1,1754 @@
...
+}
+
+/// @brief Computes velocities in x and y-dir based on momentum in x and y-dir
+/// @param N Size of the array
+/// @param tiny_h Threshold value for height
+/// @param h Height
+/// @param hu Momentum in x-dir
+/// @param hv Momentum in y-dir
+/// @param u Velocity in x-dir
+/// @param v Velocit in y-dir
```
Comment (jeff-cohere) [inline swe_roe/ex2b.c]: Velocit -> Velocity

File: swe_roe/ex2b.c [hunk 2]
```diff
@@ -0,0 +1,1754 @@
...
+  RDyCells *cells = &mesh->cells;
+  RDyEdges *edges = &mesh->edges;
+
+  // Get pointers to vector data
+  PetscScalar *x_ptr, *f_ptr, *b_ptr;
+  PetscCall(VecGetArray(app->localX, &x_ptr));
+  PetscCall(VecGetArray(F, &f_ptr));
+  PetscCall(VecGetArray(app->localB, &b_ptr));
+
+  PetscInt  dof = 3;
```
Comment (jeff-cohere) [inline swe_roe/ex2b.c]: Consider `ndof` for clarity

**Review by jeff-cohere — APPROVED** [review]
Looks good to me. So this is an example of breaking the boundary/interior flux calculations into two separate loops, and it seems pretty clean, right?

File: swe_roe/ex2b.c [hunk 3]
```diff
@@ -0,0 +1,1754 @@
...
+
+    PetscReal hl = x_ptr[l * ndof + 0];
+    PetscReal hr = x_ptr[r * ndof + 0];
+    PetscReal bl = b_ptr[l];
+    PetscReal br = b_ptr[r];
+
+    *amax_value = fmax(*amax_value, amax_vec_int[ii]);
+
+    if (bl == 0 && br == 0) {
+      // Both, left and right cells are not boundary walls
```
Comment (xdongh) [inline swe_roe/ex2b.c]: Do you mean both left and right are not ghost cells?

File: swe_roe/ex2b.c [hunk 4]
```diff
@@ -0,0 +1,1754 @@
...
+  PetscCall(TSSetExactFinalTime(ts, TS_EXACTFINALTIME_STEPOVER));
+  PetscCall(TSSetSolution(ts, X));
+  PetscCall(TSSetTimeStep(ts, app->dt));
+
+  PetscCall(TSSetFromOptions(ts));
+  PetscCall(TSSolve(ts, X));
+
+  if (app->save) {
+    char fname[PETSC_MAX_PATH_LEN];
+    sprintf(fname, "outputs/ex1_Nx_%d_Ny_%d_dt_%f_%d_np%d.dat", app->Nx, app->Ny, app->dt, app->Nt, app->comm_size);
```
Comment (xdongh) [inline swe_roe/ex2b.c]: Small typo. Should be ex2b. How about create a variable for the example name?

**Review by xdongh — COMMENTED** [review]
I only have minor comments.

**Review by donghuix — APPROVED** [review]
I left comments from another account.

File: swe_roe/ex2b.c (hunk 3 above)
Comment (bishtgautam) [reply swe_roe/ex2b.c]: No. The code enters the loop if both left and right are not reflective boundary walls.

Comment (jeff-cohere) [reply swe_roe/ex2b.c]: Maybe we could use slightly more descriptive names for these, or leave a comment somewhere describing what they mean.

**Pull request merged**